### PR TITLE
Improve collision check on hparams between LightningModule and DataModule 

### LINKED ISSUE #9492: Improve collision check on hparams between LightningModule and DataModule
### Motivation

With the recent ability to log hyperparameters on the datamodule, this exception was introduced in case the keys had overlaps between the lightning module and datamodule
https://github.com/PyTorchLightning/pytorch-lightning/blob/ec828b826717cd3b5beabcb6d0cacf41b2320a98/pytorch_lightning/trainer/trainer.py#L1043-L1053

However, this check can be overly strict: if the same hparams are shared across the LightningModule and DataModule, this will result in an error. 

### Pitch

Only raise an exception if there are overlapping keys which have different values across lightning module and datamodule
```
datamodule_hparams = self.datamodule.hparams_initial
lightning_hparams = self.lightning_module.hparams_initial
colliding_keys = lightning_hparams.keys() & datamodule_hparams.keys()
if colliding_keys:
    inconsistent_keys = []
    for key in colliding_keys:
        if lightning_hparams[key] != datamodule_hparams[key]:
            inonsistent_keys.append(key)
    if len(inconsistent_keys) > 0:
        raise MisconfigurationException(
            f"Error while merging hparams: the keys {inconsistent_keys} are present "
            "in both the LightningModule's and LightningDataModule's hparams and have different values."
        )
hparams_initial = {**lightning_hparams, **datamodule_hparams}
```

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #9492

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by kaushikb11 — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by SeanNaren — APPROVED** [review]
Thanks @ananthsub!

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1046,10 +1046,16 @@ def _log_hyperparams(self):
 
                 colliding_keys = lightning_hparams.keys() & datamodule_hparams.keys()
                 if colliding_keys:
-                    raise MisconfigurationException(
-                        f"Error while merging hparams: the keys {colliding_keys} are present "
-                        "in both the LightningModule's and LightningDataModule's hparams."
-                    )
+                    inconsistent_keys = []
+                    for key in colliding_keys:
+                        if lightning_hparams[key] != datamodule_hparams[key]:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1051]: here we assume that these objects can be compared with ==,  that's probably too strong.

For example, if these are tensors then the result of the comparison is again a tensor and we get an error. Or if these tensors don't even have the same shape, we yet get another error.

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1046,10 +1046,16 @@ def _log_hyperparams(self):
 
                 colliding_keys = lightning_hparams.keys() & datamodule_hparams.keys()
                 if colliding_keys:
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1048]: this if is redundant now

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1046,10 +1046,16 @@ def _log_hyperparams(self):
...
-                    raise MisconfigurationException(
-                        f"Error while merging hparams: the keys {colliding_keys} are present "
-                        "in both the LightningModule's and LightningDataModule's hparams."
-                    )
+                    inconsistent_keys = []
+                    for key in colliding_keys:
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1050]: No need for the variable anymore

```suggestion
                    for key in lightning_hparams.keys() & datamodule_hparams.keys():
```

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:1051]: is there a specification around what argument types can be included in the hparams, which could then guide these equality checks? is there another form of equality check we ought to use for this comparison?

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1051]: Should we use id(lightning_hparams[key]) != id(datamodule_hparams[key]) ?

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1051]: > is there a specification around what argument types can be included in the hparams


not really afaik. there are two functionalities:
1) saving hparams to checkpoint
2) logging hparams

In 1) all we require is that the object is picklable. 
In 2) it is ultimately the logger defining what it can log and what it can't log.

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py]: @awaelchli I updated to use `id` in case this would be safer. i'm not sure if this introduces issues around lifetimes or hparams potentially being updated in between when save_hyperparameters is called and when it's logged here though.

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: sounds good. hopefully nobody is using ints as keys hehe. See this [funny business here](https://github.com/satwikkansal/wtfpython#-how-not-to-use-is-operator). Where `id(257) != id(257)` but `id(256) == id(256)` 🤣

File: CHANGELOG.md [hunk 4]
```diff
@@ -196,6 +196,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Executing the `optimizer_closure` is now required when overriding the `optimizer_step` hook ([#9360](https://github.com/PyTorchLightning/pytorch-lightning/pull/9360))
 
 
+- Pass init args to `ShardedDataParallel` ([#9483](https://github.com/PyTorchLightning/pytorch-lightning/pull/9483))
+
+
```
Comment (awaelchli) [inline CHANGELOG.md:201]:
```suggestion
```

Comment (awaelchli) [reply CHANGELOG.md:201]: bad merge

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py]: If the problem is just tensors, we can use id only for them, instead of id for everything

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py]: @carmocca @awaelchli I added special asing for tensors to use the id check. LMK if you think this is good to merge now, or if there's a more general approach

File: tests/loggers/test_base.py [hunk 5]
```diff
@@ -288,3 +291,77 @@ def __init__(self, param_one, param_two):
...
+
+    class TestDataModule(BoringDataModule):
+        def __init__(self, hparams: Dict[str, Any]) -> None:
+            super().__init__()
+            self.save_hyperparameters(hparams)
+
+    class _Test:
+        ...
+
+    same_params = {1: 1, "2": 2, "three": 3.0, "test": _Test(), "4": torch.Tensor(4)}
```
Comment (carmocca) [inline tests/loggers/test_base.py:311]:
```suggestion
    same_params = {1: 1, "2": 2, "three": 3.0, "test": _Test(), "4": torch.tensor(4)}
```

**Review by carmocca — APPROVED** [review]
